proposals/25-ipld-in-ipfs.md

```diff
@@ -0,0 +1,148 @@
+Now the same query gets back `QMGHIJKL` -- a CID to download the image above of a space shuttle launch!
+
+#### Impact
+
+We have a number of existing users who have structured data they want to put in IPFS.
+Currently, they are expending effort to take that data and put it in flat files before importing into IPFS, with almost no query capabilities. 
+
+#### Leverage
+
+This a project whose greatest impact may be unblocking other projects. Among them:
+
+* Faster data transfer in IPFS (we can integrate bitswap and graphsync to speed up transfer, we cannot in the current stack)
+* Unlocking greater capabilities in the stack through use of selectors, including to-be-written selectors
+* IPLD based databases and graph querying as optimization mechanism
+* Standardizing IPLD across the stack including Filecoin (see meta doc above)
+* Coherent IPFS and Filecoin retrieval patterns (also needs Filecoin retrieval improvements)
+* Almost any effort to support dApps that work with non-static or non-file data
+
+#### Confidence
+
+On the scale: 1 -- Low
+
+Impact: 1 -- We have anecdotal examples of users who are frustrated with their inability to put semantic data into IPFS (todo: collect examples from Eric Myhre, Adin Schmahmann)
+
+Impact 3: Top user request: https://github.com/ipfs/go-ipfs/issues/7909
+
+## Project definition
+#### Brief plan of attack
+
+go-IPFS primarily uses legacy IPLD libraries -- go-ipld-format -- in the context of working primarily with UnixFS v1.
+Use of IPLD primarily flows through the [`DAGService`](https://github.com/ipfs/go-ipld-format/blob/master/merkledag.go), which sits on the BlockService, which is hardcoded to fetch through Bitswap, among other things.
+
+Several things sit on top of `DAGService` -- most prominently UnixFS.
+
+The first goal here is to get to a go-ipld-prime based `DAGService`, as close in interface to the existing DAGService, so that libraries dependent on this service would require only minimal changes.
+
+A first step is proposing an ipld-prime interface for this revamp’d DAG service, and to write a minimal scaffold translating from that interface to the current implementation.
+
+At that point, two people will continue working within the `dag service -> block service -> bitswap` and plumb through IPLD in that direction, while others will add more IPLD commands to the IPFS command line and handle propagating it to the various top level interfaces.
+
+We can also decide if we want to pursue updating some services that sit on top of `DAGService`.
+
+#### What does done look like?
+
+* P0: I can build IPFS without go-ipld-format
+* P1: I can import structured data (following links of generic codecs) directly from the IPFS command line
+* P1: structured data is importable and can be queried over the IPFS HTTP API 
+* P2: I can query structured data directly from the IPFS command line
```

Comment:
Why wouldn't this be a P1 as well?  Without this, wouldn't it mean we can import structured data but can't query it?

Reply:
In fairness these are more prioritizations against the time allotted, which is generally how I think of prioritization -- the priority is defined by its relative priorization to other tasks. If we get the P0 done, then we do the P1, then the P2. That's probably not the best way to go about it.

Reply:
Having it accessible programmatically in structured ways is probably the implicit p1 while passing in and out of the inherently serialized command line is perhaps less used by _developers_ building on IPFS?